MySQLdb/connections.py

```diff
@@ -298,6 +313,14 @@ def set_character_set(self, charset):
         super().set_character_set(charset)
         self.encoding = _charset_to_encoding.get(charset, charset)
 
+    def set_character_set_collation(self, charset, collation):
+        """Set the connection character set and collation. Use this as
+        an alternative to ``set_character_set``.
+        """
+        self.query("SET NAMES %s COLLATE %s" % (charset, collation))
```

Comment:
```suggestion
        super().set_character_set(charset)
        self.query("SET NAMES %s COLLATE %s" % (charset, collation))
```

Reply:
Note that `super().set_character_set()` not only sending `SET NAMES charset`, but also updating `MYSQL->charset` in client C library.

This means you can not avoid sending two queries (e.g. `SET NAMES charaset` and `SET NAMES charaset COLLATE collation`).
It seems weired, but it is MySQL client library design. It is hard to use in safe.

Reply:
Good point. It might be possible to avoid sending two queries. From looking at the C connector source code, it seems to be sending SET NAMES, then upon success setting the charset on the MYSQL struct. I wonder if we can replicate that functionality?

https://github.com/mariadb-corporation/mariadb-connector-c/blob/4de47e64a53a2844e3a66c2d1173ba8e1c54c8f6/libmariadb/mariadb_lib.c#L4250-L4259

Reply:
https://github.com/search?q=repo%3Amysql%2Fmysql-server%20mysql_find_charset_name&type=code

Reply:
We can use only APIs implemented by both of MariaDB and MySQL. (including old versions)

Reply:
The MySQL implementation is a bit more complicated, but the logic is similar: https://github.com/mariadb-corporation/mariadb-connector-c/blob/4de47e64a53a2844e3a66c2d1173ba8e1c54c8f6/libmariadb/mariadb_lib.c#L4250-L4259

However the APIs do seem too different. I guess we will have to settle for sending two queries. I will update the patch.